altair/utils/core.py

```diff
@@ -602,6 +620,41 @@ def parse_shorthand(
     return attrs
 
 
+def infer_vegalite_type_for_dfi_column(
+    column: Column,
+) -> Union[_InferredVegaLiteType, Tuple[_InferredVegaLiteType, list]]:
+    from pyarrow.interchange.from_dataframe import column_to_array
+
+    try:
+        kind = column.dtype[0]
+    except NotImplementedError as e:
+        # Edge case hack:
+        # dtype access fails for pandas column with datetime64[ns] type,
+        # but all we need to know is that its temporal, so check the
+        # error message for the presence of datetime64.
+        if "datetime64" in e.args[0]:
+            return "temporal"
+        raise e
+
+    if (
+        kind == DtypeKind.CATEGORICAL
+        and column.describe_categorical["is_ordered"]
+        and column.describe_categorical["categories"] is not None
+    ):
+        # Treat ordered categorical column as Vega-Lite ordinal
```

Comment:
> Or would this become difficult to go back to the [fallback](https://github.com/altair-viz/altair/pull/3114/files#diff-d110e8643c6feff37c22c1a5c7d213b8384177718b8e26edd8e1b8f2ed027a0aR591) of pandas?

Yeah, I used `pyarrow_available()` up above to make it easier to know whether to use the fallback logic. An alternative would be to raise a special exception if we get to this point and pyarrow is not available, and then use the fallback logic if that exception is raised. But this seemed more complicated than it was worth to me.

You do make a good point that this approach will result in pyarrow always being imported.  This does introduce a small one-time delay. But on the other hand, if our future direction is to rely on pyarrow and the dataframe interchange protocol instead of pandas, then we'll end up importing pyarrow at some point.